### PR TITLE
{Core} Refactor dynamic extension install in `_check_value`

### PR DESCRIPTION
**Pull request by fengzhou-msft**
**Description**

Move the logic of  dynamic extension install to a new fucntion `_check_value_in_extensions`. It will exit if the parser error is caused by an extension not installed, and simplified error handling code below by removing the `not caused_by_extension_not_installed` check based on this change.

This PR also adds the check whether the latest version of the extension is already installed to avoid infinite loop in cases like https://github.com/Azure/azure-cli/issues/17497

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Review by houk-ms — APPROVED** [review]
good job

File: src/azure-cli-core/azure/cli/core/parser.py [hunk 1]
```diff
@@ -392,135 +391,165 @@ def _get_extension_run_after_dynamic_install_config(self):
                                                                   default_value) if cli_ctx else default_value
         return run_after_extension_installed
 
-    def _check_value(self, action, value):  # pylint: disable=too-many-statements, too-many-locals, too-many-branches
+    def _check_value_in_extensions(self, cli_ctx, args, no_prompt):  # pylint: disable=too-many-statements, too-many-locals
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/parser.py:394]: Can we move this function to a separate file?

File: src/azure-cli-core/azure/cli/core/parser.py [hunk 2]
```diff
@@ -392,135 +391,165 @@ def _get_extension_run_after_dynamic_install_config(self):
...
+            recommender.set_help_examples(self.get_examples(command_name_inferred))
+            recommendations = recommender.provide_recommendations()
+            if recommendations:
+                az_error.set_aladdin_recommendation(recommendations)
+
+            # remind user to check extensions if we can not find a command to recommend
+            if isinstance(az_error, CommandNotFoundError) \
+                    and not az_error.recommendations and self.prog == 'az' \
+                    and use_dynamic_install == 'no':
+                az_error.set_recommendation(EXTENSION_REFERENCE)
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/parser.py:552]: Maybe we can move these lines to another file too.

File: src/azure-cli-core/azure/cli/core/parser.py (hunk 1 above)
Comment (fengzhou-msft) [reply src/azure-cli-core/azure/cli/core/parser.py]: Done

**Review by msyyc — APPROVED** [review]

**Review by jiasli — APPROVED** [review]

**Pull request merged**